app/controllers/api/v1x0/icons_controller.rb

```diff
@@ -4,16 +4,47 @@ class IconsController < ApplicationController
       include Api::V1x0::Mixins::IndexMixin
 
       def show
-        portfolio_item = PortfolioItem.find_by!(:id => params.require(:portfolio_item_id))
-        raise ActiveRecord::RecordNotFound, "Icon not present on Portfolio Item" if portfolio_item.service_offering_icon_ref.nil?
+        render :json => Icon.find(params.require(:id))
+      end
+
+      def create
+        icon = Icon.create!(icon_params)
+        render :json => icon
+      end
 
-        so = ServiceOffering::Icons.new(portfolio_item.service_offering_icon_ref)
-        icon = so.process.icon
+      def destroy
+        Icon.find(params.require(:id)).destroy
+        head :no_content
+      end
+
+      def update
+        icon = Icon.find(params.require(:id))
+        icon.update!(icon_patch_params)
+
+        render :json => icon
+      end
 
+      def raw_icon
+        icon = find_icon
         send_data(icon.data,
                   :type        => MimeMagic.by_magic(icon.data).type,
                   :disposition => 'inline')
       end
+
+      private
+
+      def icon_params
+        params.require(:data)
+        icon_patch_params
+      end
+
+      def icon_patch_params
+        params.permit(:data, :source_ref, :source_id, :portfolio_item_id)
+      end
+
+      def find_icon
+        params[:icon_id].present? ? Icon.find(params[:icon_id]) : Icon.find_by!(:portfolio_item_id => params[:portfolio_item_id])
```

Comment:
Do we run into any problems here since we're not `.permit`-ing or `.require`-ing the parameters that we're trying to find the icon by? Maybe this should be `params.permit(:icon_id).present?` and then `params.permit(:portfolio_item_id)`?

The specs still seem to work so I'm not sure if we care because we're not doing an `update` or a `create` where someone could pass problematic parameters anyway, but we still do `.require(:id)` up on line 7 for the `.find` as well.

Reply:
Good catch, I added a parameter to the `find_icon` method and require'd the `*_id` parameters.